jhu-idc/idc-isle-buildkit
# Dead letter queue microservice

## About

This PR introduces a new buildkit image named `dlq`, which is intended to be used as a microservice for reading undeliverable messages off of the ActiveMQ dead letter queue, and emitting the message to `stderr`.

This PR updates the alpaca image to direct undelivered messages to the ActiveMQ dead letter queue, rather than logging them to the console.

## DLQ

The image comes pre-configured and ought to just work™️ when invoked as a docker container.  The README contains the details.

If `DRUPAL_JWT_PUBLIC_KEY` is provided as an environment variable, then the JWTs associated with each message will be validated.  This is an optional value, and does not need to be set unless there is particular concern that JWTs may be insecure or spoofed.

## Alpaca

A new environment variable `ALPACA_DLQ` is introduced with a default value of `broker:queue:ActiveMQ.DLQ`.  The value is a Camel endpoint URI which accepts undeliverable messages. 

## Testing
* Update idc-isle-dc as follows:
    * Modify `TAG` in `.env` to `upstream-20200824-f8d1e8e-67-g146a945` 
    * Modify `docker-compose.alpaca.yml`
        * Add the following key/value pairs to the environment (this will speed up your testing):
            * `ALPACA_REDELIVERIES: 1`
            * `ALPACA_REDELIVERYDELAY: 5000`
    * Modify the key `HOMARUS_JWT_PUBLIC_KEY` in `docker-compose.env.yml` by changing the last line of the key value from `rwIDAQAB` to `rwIDAQAA`, rendering the key invalid.
* Reset idc-isle-dc by invoking `docker-compose down -v`
* Regenerate `docker-compose.yml`: `make -B docker-compose.yml`
* Launch IDC: `make up`
> If `make up` fails, you may need to invoke `docker-compose pull` then retry `make up`
* After IDC has started, verify Alpaca will route messages to the dead letter queue:
```shell
$ docker-compose exec alpaca grep dlq deploy/ca.islandora.alpaca.connector.homarus.blueprint.xml
       <cm:property name="error.dlq" value="broker:queue:ActiveMQ.DLQ"/>
```

Ok, now we're ready to party.

Start the DLQ microservice by running:
```shell
docker run --rm --network idc_default ghcr.io/jhu-sheridan-libraries/idc-isle-dc/dlq:upstream-20200824-f8d1e8e-67-g146a945 -queue ActiveMQ.DLQ -host activemq -pass password -config ./config.json
```

You should see output like:
```
2021/12/06 17:28:27 Configuring stomp-logger *listener.StompLoggerHandler
2021/12/06 17:28:27 activating handler: stomp-logger
2021/12/06 17:28:27 Configuring jwt-logger *listener.JWTLoggingHandler
2021/12/06 17:28:27 activating handler: jwt-logger
2021/12/06 17:28:27 listener: successfully connected to activemq:61613, reading messages from ActiveMQ.DLQ
```

Next, you're going to run a migration using videos (so that homarus is used).  Because you munged the value of `HOMARUS_JWT_PUBLIC_KEY`, requests from Alpaca to homarus will fail to authenticate, and Alpaca will deliver the message to dead letter queue.  You can perform a migration using whatever means you're comfortable with, but you must perform a video migration, otherwise no messages will be in error, and nothing will route to the DLQ.
> Don't forget to start the migration-assets container

After performing the video migration, you should see some output in the Alpaca console that indicates failures to invoke the Homarus endpoint.  After attempting to redeliver the messages, they should be delivered to the ActiveMQ dead letter queue.  This may take five to ten minutes.  As the messages are delivered to the DLQ, you should see output from the dead letter microservice, something like:

```
2021/12/06 16:50:37 [StompLoggerHandler] [ID:8ae28c1ed801-39917-1638803056166-5:6:-1:1:2] STOMP headers and body
  Content Type: 
  Destination: /queue/ActiveMQ.DLQ
  Subscription: 1
  Headers:
    expires: 1639414237168
    scheduledJobId: ID:8ae28c1ed801-39917-1638803056166-5:6:-1:1:5
    dlqDeliveryFailureCause: java.lang.Throwable: Delivery[7] exceeds redelivery policy limit:RedeliveryPolicy {destination = null, collisionAvoidanceFactor = 0.15, maximumRedeliveries = 6, maximumRedeliveryDelay = -1, initialRedeliveryDelay = 1000, useCollisionAvoidance = false, useExponentialBackOff = false, backOffMultiplier = 5.0, redeliveryDelay = 1000, preDispatchCheck = true}, cause:null
    destination: /queue/ActiveMQ.DLQ
    ack: ID:8ae28c1ed801-39917-1638803056166-6:8
    subscription: 1
    priority: 4
    Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJpYXQiOjE2Mzg4MDkyNTcsImV4cCI6MTYzODgyMzY1Nywid2ViaWQiOiIxIiwiaXNzIjoiaHR0cHM6XC9cL2lzbGFuZG9yYS1pZGMudHJhZWZpay5tZSIsInN1YiI6ImFkbWluIiwicm9sZXMiOlsiYXV0aGVudGljYXRlZCJdfQ.EQrmBKu9Bl_fufJwhtzYtfsQJKPmei1McdB4opJ_wRtMfCieysGXJzTQErIxvAsyvTreTfW0wu15zJhwKZ5jpukkQBfIM3YlmRDY57rfcDR6QeSIgWZs-H3fRnZOP0l0hV-4rbBvER9WSI6BpXl5Bpw3n2_m03CTG1I1q94TfAqp4zMw4L6EcQYPUyGdFtZa5uu0xYWkShnf_aRzPXienuIiO--Pk20nDU16EgHtWuLssq9d4CbcJxeA9hz-aRhkVTf8LTgIppw-x-lwxEtWO97ORK_eriXBM2nmQ417QnWlfopctmEfKfFkRxs8uw14WSepnujjnG7iPKMGvrcm8A
    original-destination: /queue/islandora-connector-homarus
    message-id: ID:8ae28c1ed801-39917-1638803056166-5:6:-1:1:2
    redeliveryDelay: 60000
    persistent: true
    timestamp: 1638809257415
  Body:
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "actor": {
    "id": "urn:uuid:8b2345a3-20f6-4d27-bc72-b4b2a6189498",
    "type": "Person",
    "url": [
      {
        "href": "http://islandora-idc.traefik.me/user/1",
        "mediaType": "text/html",
        "name": "Canonical",
        "rel": "canonical",
        "type": "Link"
      }
    ]
  },
  "attachment": {
    "content": {
      "args": "-ss 00:00:01.000 -frames 1 -vf scale=100:-2",
      "destination_uri": "https://islandora-idc.traefik.me/node/39/media/image/20",
      "file_upload_uri": "private://2021-12/39-Thumbnail Image.jpg",
      "mimetype": "image/jpeg",
      "source_uri": "http://islandora-idc.traefik.me/system/files/2021-12-06/smptebars_2.mp4"
    },
    "mediaType": "application/json",
    "type": "Object"
  },
  "object": {
    "id": "urn:uuid:f3480f98-d11e-439a-a8c4-83d1baf5df93",
    "url": [
      {
        "href": "http://islandora-idc.traefik.me/node/39",
        "mediaType": "text/html",
        "name": "Canonical",
        "rel": "canonical",
        "type": "Link"
      },
      {
        "href": "http://islandora-idc.traefik.me/node/39?_format=json",
        "mediaType": "application/json",
        "name": "JSON",
        "rel": "alternate",
        "type": "Link"
      },
      {
        "href": "http://islandora-idc.traefik.me/node/39?_format=jsonld",
        "mediaType": "application/ld+json",
        "name": "JSONLD",
        "rel": "alternate",
        "type": "Link"
      }
    ]
  },
  "summary": "Generate Derivative",
  "target": "http://fcrepo.isle-dc.localhost/fcrepo/rest/",
  "type": "Activity"
}
2021/12/06 16:50:37 [JWTLoggingHandler] [ID:8ae28c1ed801-39917-1638803056166-5:6:-1:1:2] handler: JWT could not be verified: handler: unable to verify JWT, invalid PEM encoded RSA PUBLIC KEY
2021/12/06 16:50:37 [JWTLoggingHandler] [ID:8ae28c1ed801-39917-1638803056166-5:6:-1:1:2] handler: JWT claims:
  sub: admin
  roles: [authenticated]
  iat: 1.638809257e+09 (2021-12-06T16:47:37Z)
  exp: 1.638823657e+09 (2021-12-06T20:47:37Z)
  webid: 1
  iss: https://islandora-idc.traefik.me
```
